lms/djangoapps/grades/models.py

```diff
@@ -236,9 +241,43 @@ def bulk_get_or_create(cls, block_record_lists, course_key):
         BlockRecordList objects for the given course_key, but
         only for those that aren't already created.
         """
-        existent_records = {record.hashed: record for record in cls.bulk_read(course_key)}
+        existent_records = cls.bulk_read(course_key)
         non_existent_brls = {brl for brl in block_record_lists if brl.hash_value not in existent_records}
-        cls.bulk_create(non_existent_brls)
+        created = cls.bulk_create(non_existent_brls)
+        cls.update_cache(course_key, created)
+
+
+    @classmethod
+    def prefetch(cls, course_key):
+        """
+        Prefetches visible blocks for the given course.
+        """
+        prefetched = {record.hashed: record for record in cls.objects.filter(course_id=course_key)}
+        get_cache(cls.CACHE_NAMESPACE)[cls._cache_key(course_key)] = prefetched
+        return prefetched
+
+    @classmethod
+    def update_cache(cls, course_key, visible_blocks):
+        """
+        Adds a specific set of visible blocks to the request cache.
+        This assumes that prefetch has already been called.
+        """
+        get_cache(cls.CACHE_NAMESPACE)[cls._cache_key(course_key)].update(
+            {visible_block.hashed: visible_block for visible_block in visible_blocks}
+        )
+
+    @classmethod
+    def clear_cache(cls, course_key):
+        """
+        Clears the cache of all contents for a given course.
+        """
+        cache = get_cache(cls.CACHE_NAMESPACE)
+        if course_key in cache:
+            del cache[cls._cache_key(course_key)]
```

Comment:
Nit: You can replace the above 2 lines with: `cache.pop(cls._cache_key(course_key), None)`
So you don't have to explicitly query membership in the `dict`.